src/lib.rs

```diff
@@ -1093,6 +1100,46 @@ impl<K, V, S> Drop for FlurryHashMap<K, V, S> {
     }
 }
 
+impl<K, V, S> Extend<(K, V)> for &FlurryHashMap<K, V, S>
+where
+    K: Sync + Send + Clone + Hash + Eq,
+    V: Sync + Send,
+    S: BuildHasher,
+{
+    // TODO: Implement Java's `tryPresize` method to pre-allocate space for
+    // the incoming entries
+    fn extend<T: IntoIterator<Item = (K, V)>>(&mut self, iter: T) {
+        let guard = crossbeam::epoch::pin();
+
+        for (key, value) in iter {
+            (*self).put(key, value, false, &guard);
+        }
+    }
+}
+
+impl<K, V> FromIterator<(K, V)> for FlurryHashMap<K, V, RandomState>
+where
+    K: Sync + Send + Clone + Hash + Eq,
+    V: Sync + Send,
+{
+    fn from_iter<T: IntoIterator<Item = (K, V)>>(iter: T) -> Self {
+        let mut it = iter.into_iter();
+
+        let map = if let Some((key, value)) = it.next() {
+            let guard = crossbeam::epoch::pin();
```

Comment:
Hmm, one concern here is that we are now using _two_ guards in order to make the map: one to do the first `put`, and one that is created inside `extend`. I think we'll probably want a (private) method
```rust
fn put_all<I: Iterator>(&self, iter: I, guard: &Guard)
```
that both the `Extend` and `FromIterator` methods call. `Extend` can make its own guard like it currently does, and here we can create a `epoch::unprotected` and use that for both the `put` and then subsequent `put_all`.

Reply:
Indeed. For now this looks like the best solution possible, but if we were to implement `tryPrecize`, we could do something similar to what's done in [`HashMap::from_iter`](https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#2400-2410), moving the code in `put_all` to the `Extend` implementation and leveraging it in the `FromIterator` implementation.

What do you think?

Reply:
Hmm, yes, that would probably work. I think ultimately we'll likely end up with something like [`HashMap::extend`](https://github.com/rust-lang/hashbrown/blob/6f5826ed934a5e6342681c27618bea71f9b0c800/src/map.rs#L2643-L2658) (plus the `.next()` optimization from above).

We _may_ still want to special-case `FromIterator`, since it can use `with_capacity` (which is simple) rather than `tryPresize` (which is complicated), but :man_shrugging:. We definitely want to use an unprotected guard in `FromIterator` though I feel like.

Reply:
Makes sense. Let's go with it.

Reply:
Anyway, `tryPrecize` should be implemented at some point. Unfortunatelly I don't feel like I have the level of understanding necessary for doing so.

Reply:
That's totally fine — just leave appropriate the TODOs and they'll be a learning opportunity for someone later!